ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/CommandStoreTest.java

```diff
@@ -292,14 +290,121 @@ public void shouldDistributeCommand() throws ExecutionException, InterruptedExce
     assertThat(record.value().getOriginalProperties(), equalTo(ksqlConfig.getAllConfigPropsWithSecretsObfuscated()));
   }
 
+  @Test
+  public void shouldIncludeTopicOffsetInSuccessfulQueuedCommandStatus()
+      throws InterruptedException, ExecutionException {
+    // Given:
+    final CommandStore commandStore = createCommandStoreThatAssignsSameId(commandId);
+
+    expect(commandProducer.send(anyObject(ProducerRecord.class))).andReturn(future);
+    expect(future.get()).andReturn(recordMetadata);
+    replay(commandProducer, future);
+
+    // When:
+    final QueuedCommandStatus commandStatus =
+        commandStore.enqueueCommand(statementText, statement, KSQL_CONFIG, OVERRIDE_PROPERTIES);
+
+    // Then:
+    assertThat(commandStatus.getCommandOffset(), equalTo(recordMetadata.offset()));
+
+    verify(commandProducer, future);
+  }
+
+  @Test
+  public void shouldReturnCompletedFutureIfOffsetReached() {
+    // Given:
+    givenCmdConsumerAtPosition(1);
+
+    final CommandStore commandStore = createCommandStore();
+
+    // When:
+    final CompletableFuture<Void> future = commandStore.getConsumerPositionFuture(0);
+
+    // Then:
+    assertFutureIsCompleted(future);
+    verify(commandConsumer);
+  }
+
+  @Test
+  public void shouldReturnUncompletedFutureIfOffsetNotReached() {
+    // Given:
+    givenCmdConsumerAtPosition(2);
+
+    final CommandStore commandStore = createCommandStore();
+
+    // When:
+    final CompletableFuture<Void> future = commandStore.getConsumerPositionFuture(2);
+
+    // Then:
+    assertFutureIsNotCompleted(future);
+    verify(commandConsumer);
+  }
+
+  @Test
+  public void shouldCompleteFutureWhenOffsetIsReached() {
+    // Given:
+    final CommandStore commandStore = createCommandStore();
+
+    givenCmdConsumerAtPosition(0);
+    final CompletableFuture future = commandStore.getConsumerPositionFuture(2);
+    givenCmdConsumerAtPosition(3, true);
+
+    // When:
+    commandStore.getNewCommands();
+
+    // Then:
```

Comment:
Ideally we'd also like to verify that the completed future is removed from `commandStore.commandOffsetFutures`, but as-implemented that's hard to check because `commandOffsetFutures` is a private member variable. @rodesai suggested potentially refactoring the CompletableFuture code added by this PR out of CommandStore and into a separate interface used by CommandStore, but I wanted to get feedback on the general approach before making the refactor. Thoughts?

Reply:
Yeah, pulling it out into a `OffsetFutureStore` or something would make it more testable.  If you go down my route of making it a map as well it would mean you can have a separate test for the store, covering all the corner cases, etc.